source/application/AMPIC887DriverServer/AMPIC887Controller.cpp

```diff
@@ -1,28 +1,75 @@
 #include "AMPIC887Controller.h"
 #include "PI_GCS2_DLL.h"
+#include "GCS2Commands/AMGCS2AsyncCommand.h"
+#include "GCS2Commands/AMGCS2InitializeControllerStateCommand.h"
+
+#include "GCS2Commands/AMGCS2SetCommandLevelCommand.h"
+#include "GCS2Commands/AMGCS2SetCycleTimeCommand.h"
+#include "GCS2Commands/AMGCS2SetRecordTriggerSourceCommand.h"
+#include "GCS2Commands/AMGCS2SetDataRecorderConfigurationCommand.h"
+#include "GCS2Commands/AMGCS2SetLowSoftLimitsCommand.h"
+#include "GCS2Commands/AMGCS2SetHighSoftLimitsCommand.h"
+#include "GCS2Commands/AMGCS2SetPivotPointCommand.h"
+#include "GCS2Commands/AMGCS2SetSoftLimitsStatusCommand.h"
+#include "GCS2Commands/AMGCS2SetStepSizeCommand.h"
+#include "GCS2Commands/AMGCS2SetSystemVelocityCommand.h"
+#include "GCS2Commands/AMGCS2SetServoModeCommand.h"
+
+#include "GCS2Commands/AMGCS2GetCurrentPositionCommand.h"
+#include "GCS2Commands/AMGCS2AsyncMoveCommand.h"
+#include "GCS2Commands/AMGCS2AsyncMoveRelativeCommand.h"
+#include "GCS2Commands/AMGCS2AsyncReferenceMoveCommand.h"
+
+#include "GCS2Commands/AMGCS2StopCommand.h"
+#include "GCS2Commands/AMGCS2HaltSmoothlyCommand.h"
+
+#include "AMPIC887DataRecorderConfiguration.h"
+
+#include "AMGCS2Support.h"
 
 AMPIC887Controller::AMPIC887Controller(const QString& name, const QString& hostname)
+	:QObject(0)
 {
 	hostname_ = hostname;
 	name_ = name;
-	connect();
+	connectToController();
+	isBusy_ = false;
+	currentPositionRefreshRequired_ = false;
+	controllerState_ = new AMPIC887ControllerState();
+	xMotions_ = 0;
+	yMotions_ = 0;
+	zMotions_ = 0;
+	uMotions_ = 0;
+	vMotions_ = 0;
+	wMotions_ = 0;
+	if(connectionEstablished()) {
+		initializeControllerStateData();
+	} else {
+		lastError_ = "No connection obtained with the controller";
+	}
 }
 
-void AMPIC887Controller::runCommand(AMGCS2Command* command)
+AMPIC887Controller::~AMPIC887Controller()
 {
-	if(command) {
-		command->setControllerId(id_);
-		command->run();
+	if(controllerState_ != 0) {
+		delete controllerState_;
 	}
 }
 
-bool AMPIC887Controller::connect()
+QString AMPIC887Controller::lastError() const
+{
+	return lastError_;
+}
+
+bool AMPIC887Controller::connectToController()
 {
 	id_ = PI_ConnectTCPIP(hostname_.toStdString().c_str(), CONTROLLER_PORT);
-	return connectionEstablished();
+	bool success = connectionEstablished();
```

Comment:
Unless there is more to come, it would be simpler for `return connectionEstablished()'`.